src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java

```diff
@@ -68,5 +69,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public abstract Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException;
```

Comment:
The reason for the somewhat bulky `parseContext.indexQueryParserService().queryParser(BoolQueryParser.NAME)` is that some of these Parser have other stuff injected (e.g. Settings). This is why (2) would be a bit difficult for them. I'd tend to (1) then. We could have some `final static String PARSER_NAME` in each Builder class and then do the connection in a default impl.